src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -20,75 +23,170 @@ internal static class HPackEncoder
         //          So for now, don't do dynamic encoding.
 
         public static bool EncodeHeader(string name, string value, Span<byte> buffer, out int length)
+        {
+            Debug.Assert(value != null);
+            return EncodeHeader(name, value, null, null, buffer, out length);
+        }
+
+        public static bool EncodeHeader(Headers.HeaderDescriptor header, string value, Span<byte> buffer, out int length)
+        {
+            Debug.Assert(value != null);
+            return EncodeHeader(header.Name, value, null, null, buffer, out length);
+        }
+
+        public static bool EncodeHeader(Headers.HeaderDescriptor header, string[] values, string separator, Span<byte> buffer, out int length)
+        {
+            Debug.Assert(values != null);
+            Debug.Assert(values.Length > 0);
+            return EncodeHeader(header.Name, values, separator, buffer, out length);
+        }
+
+        public static bool EncodeHeader(string name, string[] values, string separator, Span<byte> buffer, out int length)
+        {
+            Debug.Assert(values != null);
+            Debug.Assert(values.Length > 0);
+            if (values.Length == 1)
+            {
+                return EncodeHeader(name, values[0], null, null, buffer, out length);
+            }
+
+            // When we have more values, separator must be provided.
+            Debug.Assert(separator != null);
+            return EncodeHeader(name, null, values, separator, buffer, out length);
+        }
+
+        private static bool EncodeHeader(string name, string value, string[] values, string separator, Span<byte> buffer, out int length)
         {
             int i = 0;
             length = 0;
 
-            if (buffer.Length == 0)
+            // We need at least \0 and twice length plus one octet string
```

Comment:
Is this actually true? I think that having an empty (zero-length) value is allowed.

Reply:
I was looking for the answer and it seems like older spec did allow it but current specification does not. 
In worst case we will ask for buffer than needed if this really fits on buffer size boundary, right?

Reply:
Where does the RFC disallow this?

It doesn't really matter either way, since we will check again below. But we should understand what's allowed and what's not.

Reply:
I found discussion about on Internet. Old HTTP/1.1 allowed empty and NULL values. I also tested zero length value headers again nginx and Azure IIS servers and they both happily take them. 

I updated code accordingly.